### PR TITLE
fix #39379, use first∘LinearIndices instead of 1

### LINKED ISSUE #39379: zero-argument getindex method in Base incorrectly assumes 1-based indexing
It seems that the bounds check used in `Base` for zero-dimensional indexing hardcodes a comparison to `1`, assuming one-based indexing.

This means that an array with nonstandard axes can successfully pass the bounds check and then access out-of-bounds memory when indexing. I encountered this in an [issue](https://github.com/JuliaArrays/OffsetArrays.jl/issues/194) on OffsetArrays and @jishnub tracked it down to this line in `Base`:

https://github.com/JuliaLang/julia/blob/69d24532980bda6db510b6d4b71f66e7ffff2e81/base/abstractarray.jl#L642

### PR DESCRIPTION
**Pull request by mbauman**
to resolve a zero-index getindex to a linear index.

fix #39379

### CODE REVIEW

**Comment by mbauman** [comment]
Note that this is a zero-cost abstraction:

```
julia> @code_llvm Base._to_linear_index([])
;  @ abstractarray.jl:1196 within `_to_linear_index'
define i64 @julia__to_linear_index_133({}* nonnull align 16 dereferenceable(40) %0) {
top:
  ret i64 1
}
```

File: test/offsetarray.jl [hunk 1]
```diff
@@ -715,6 +715,16 @@ end
     @test axes(S) == (OffsetArrays.IdOffsetRange(0:1), Base.OneTo(2), OffsetArrays.IdOffsetRange(2:5))
 end
 
+@testset "Zero-index indexing" begin
+    @test OffsetArray([6], 2:2)[] == 6
+    @test OffsetArray(fill(6, 1, 1), 2:2, 3:3)[] == 6
+    @test OffsetArray(fill(6))[] == 6
+
```
Comment (mbauman) [inline test/offsetarray.jl:722]:
```suggestion

```

**Comment by simeonschaub** [comment]
Is it a problem that `_to_linear_index` will now throw for empty arrays? It might at least make some error messages potentially more confusing.

**Comment by simeonschaub** [comment]
Ah, never mind. We do actually support `first` for empty ranges.

**Comment by mbauman** [comment]
> Is it a problem that _to_linear_index will now throw for empty arrays?

It doesn't (see the unconditional `ret 1` above).

```julia
julia> LinearIndices([])
0-element LinearIndices{1, Tuple{Base.OneTo{Int64}}}

julia> first(LinearIndices([]))
1
```

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("array", vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/39393#issuecomment-767345688):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(`git pull -X ours`, ProcessExited(1)) [1]

```
Unfortunately, the logs could not be uploaded.
cc @christopher-dG

**Comment by christopher-dG** [comment]
I upgraded nanosoldier.jl on one of the nodes, I'll have a look at this in the morning

**Comment by mbauman** [comment]
I’d be surprised if we even have any benchmarks for `A[]` where A is not zero-dimensional.

**Comment by KristofferC** [comment]
I just wanted to check that there wasn't any overhead from this in normal benchmarks (things starting to fail to inline etc).

**Pull request merged**